### ISSUE TITLE
About finetuning LLAMA for summerization task

### ISSUE THREAD

**Issue by egg-west** [alexrame/rewardedsoups#1]
Hi all,

I was following the README in the llama folder, having run these commands
```
python3 train_ppo.py --task summary --dataset_name news --reward_models Tristan/gpt2_reward_summarization --output_folder ${folder_r1}
python3 train_ppo.py --task summary --dataset_name news --reward_models CogComp/bart-faithful-summary --dataset_name news-detector --reward_formats '1-0' --output_folder ${folder_r2}

python3 inference_rewardedsoups.py --task summary --dataset_name news --peft_names ${folder_r1} ${folder_r2}
```
The results I got are as below:
```
d[ 0.0 ] = [{'LABEL_0': 1.0334709184616804, 'n': 'grs'}, {'HALLUCINATED': 1.3255301121249794, 'FAITHFUL': -0.5046869936399162, 'n': 'bfsd'}, {'length': 200}]
d[ 0.1 ] = [{'LABEL_0': 1.0334709184616804, 'n': 'grs'}, {'HALLUCINATED': 1.3255301121249794, 'FAITHFUL': -0.5046869936399162, 'n': 'bfsd'}, {'length': 200}]
d[ 0.2 ] = [{'LABEL_0': 1.0334709184616804, 'n': 'grs'}, {'HALLUCINATED': 1.3255301121249794, 'FAITHFUL': -0.5046869936399162, 'n': 'bfsd'}, {'length': 200}]
d[ 0.3 ] = [{'LABEL_0': 1.0334709184616804, 'n': 'grs'}, {'HALLUCINATED': 1.3255301121249794, 'FAITHFUL': -0.5046869936399162, 'n': 'bfsd'}, {'length': 200}]
d[ 0.4 ] = [{'LABEL_0': 1.0334709184616804, 'n': 'grs'}, {'HALLUCINATED': 1.3255301121249794, 'FAITHFUL': -0.5046869936399162, 'n': 'bfsd'}, {'length': 200}]
d[ 0.5 ] = [{'LABEL_0': 1.0334709184616804, 'n': 'grs'}, {'HALLUCINATED': 1.3255301121249794, 'FAITHFUL': -0.5046869936399162, 'n': 'bfsd'}, {'length': 200}]
d[ 0.6 ] = [{'LABEL_0': 1.0334709184616804, 'n': 'grs'}, {'HALLUCINATED': 1.3255301121249794, 'FAITHFUL': -0.5046869936399162, 'n': 'bfsd'}, {'length': 200}]
d[ 0.7 ] = [{'LABEL_0': 1.0334709184616804, 'n': 'grs'}, {'HALLUCINATED': 1.3255301121249794, 'FAITHFUL': -0.5046869936399162, 'n': 'bfsd'}, {'length': 200}]
d[ 0.8 ] = [{'LABEL_0': 1.0334709184616804, 'n': 'grs'}, {'HALLUCINATED': 1.3255301121249794, 'FAITHFUL': -0.5046869936399162, 'n': 'bfsd'}, {'length': 200}]
d[ 0.9 ] = [{'LABEL_0': 1.0334709184616804, 'n': 'grs'}, {'HALLUCINATED': 1.3255301121249794, 'FAITHFUL': -0.5046869936399162, 'n': 'bfsd'}, {'length': 200}]
d[ 1.0 ] = [{'LABEL_0': 1.0334709184616804, 'n': 'grs'}, {'HALLUCINATED': 1.3255301121249794, 'FAITHFUL': -0.5046869936399162, 'n': 'bfsd'}, {'length': 200}]
```

May I kindly ask for hints about the reason for these results? They get the same scores with different objective weights. Was it caused by not tuning the number of epochs for fine-tuning?

Kind regards,
Ethan

**Comment by egg-west** [comment]
I hope this information helps: I observed that `objective/kl` has only negative values, for both `train_ppo.py` I ran.

**Comment by alexrame** [comment]
Hi, thanks for your question. From my understanding, it seems like your obtained scores are very close to those I obtained with the LLaMA initialization. Thus, I think you had the right intuition, and that you have not fine-tuned for sufficient steps. That would also explain why the KL is negative (as it can be for the first few steps). 

To verify this, please list the folders in the different folders, i.e., 'ls $folder_r1'. As I [save](https://github.com/alexrame/rewardedsoups/blob/d345d3c282c16085ae9a51f1072c4cb97217b2d3/llama/utils/ppo_utils.py#L179) the checkpoints when '$step %10==1', you should have multiple folders of the format ["epoch" + $step](https://github.com/alexrame/rewardedsoups/blob/d345d3c282c16085ae9a51f1072c4cb97217b2d3/llama/utils/ppo_utils.py#L243).

I hope this helps, closing in the meantine.

**Comment by egg-west** [comment]
Hi, thanks for the detailed reply. I reopen this issue for the same question but offering more details. 

I was using a single A100 80G GPU. Finetuning one epoch costs approximately two days. And the results I post are from the model fine-tuned for one epoch. May I ask:
1) whether this time (two days for one epoch) is normal?
2) how many epochs were used to obtain the results in your README?

Regards,
Ethan

**Comment by alexrame** [comment]
A full fine-tuning in two days should be normal. However, my best guess is that you did not do a single epoch, but actually a single **gradient** step. Our main interpolating results were obtained after 101 steps (roughly 2 epochs). Yet, you can observe intermediary results in Figure3a, where each points is obtained when step==1%10. If you want more info, please share what is inside your training folder; also, when launching the evaluation, you can specify which [step you want to use](https://github.com/alexrame/rewardedsoups/blob/d345d3c282c16085ae9a51f1072c4cb97217b2d3/llama/utils/inference_utils.py#L219).

**Comment by egg-west** [comment]
There are two folders for each reward model. The first one consists of `epoch1`, `epoch11`, `epoch21`, .. `epoch51`, and an `epochNone`.  

For another one: `epoch1`, `epoch11`, `epoch21`, .. `epoch41`, and an `epochNone`.  

I tried different commands for evaluation:
`python3 inference_rewardedsoups.py --task summary --dataset_name news --peft_names r0/epochNone r1/epochNone`
`python3 inference_rewardedsoups.py --task summary --dataset_name news --peft_names r0/epoch51 r1/epoch41`
Both of these commands print out similar results among preferences.

Much Thanks!

**Comment by alexrame** [comment]
I've investigated your error and was unable to reproduce it. I obtain different values when infering with different folders. Maybe the bug comes from a different package version?

To try to isolate your bug, could you please investigate whether the issue lies
- in the training: specifically, the model does not change or is not saved properly? This can be analyzed by checking whether the different folders (for different rewards and/or different steps) contain different weights.
- in the inference: specifically, there is an issue when loading the fine-tuned weights ? A minimal setup to investigate this is simply with: ```python3 inference_rewardedsoups.py --task summary --dataset_name samples --peft_names r0/epoch1``` where ```dataset_name samples```provides toy prompts, and giving only one peft_name.

PS: it does not seem correlated, but there is a bug in your training command. It should be 
```python3 train_ppo.py --task summary --dataset_name news --reward_models CogComp/bart-faithful-summary-detector --reward_formats '1-0' --output_folder ${folder_r2}```.

PS2: Also, have you fixed the negative KL issue? If not, this clearly indicates a problem training where the weights are never updated.

**Comment by egg-west** [comment]
Thanks for the detailed guidance. I tried to inference models from different epochs and found they get exactly the same answers. Then I looked into the training KL divergence and found all recorded KL divergence values are negative, which should be the reason that the weight is not updated.

![image](https://github.com/alexrame/rewardedsoups/assets/27271284/074821cd-ebea-40b5-aeef-51a5769d3084)

**Comment by alexrame** [comment]
OK thanks for the additional information. I believe you should have an inconsistency in the packages versions. I am currently writing my thesis (so don't have the time for checking it), but please try to reproduce it with the package versions enclosed [here](https://github.com/alexrame/rewardedsoups/blob/main/llama/requirements.txt).

**Comment by xyzCS** [comment]
> Thanks for the detailed guidance. I tried to inference models from different epochs and found they get exactly the same answers. Then I looked into the training KL divergence and found all recorded KL divergence values are negative, which should be the reason that the weight is not updated.
> 
> ![image](https://private-user-images.githubusercontent.com/27271284/253292240-074821cd-ebea-40b5-aeef-51a5769d3084.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTMxNzAyOTUsIm5iZiI6MTcxMzE2OTk5NSwicGF0aCI6Ii8yNzI3MTI4NC8yNTMyOTIyNDAtMDc0ODIxY2QtZWJlYS00MGI1LWFlZWYtNTFhNTc2OWQzMDg0LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA0MTUlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNDE1VDA4MzMxNVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWMyYTgzZjE4OTYwOTI1OWFlNDU2YjE2M2U2NjNlODU5OWRhYWU1ZGMxNjIwNWY3MGUzZTFhNmE0Yjc5MDFmZDAmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.VQTIO1Hjt7ZdB_Z4zLgt6Ivo4ELPaxmKhAWXUilyqxU)

Hi, Have you fixed this problem? I also encountered the same problem,  and I used the package versions provided by the author, the KL values were negative during the training (Assistant dataset). Many thanks.

**Comment by alexrame** [comment]
No I have not checked this out because of lack of additional information. Though, as much progress have been made recently in the open-source of fine-tuning softwares, I would advise patching the latest version of [trl](https://github.com/huggingface/trl) and apply weight interpolation on top of it.

**Comment by xyzCS** [comment]
> No I have not checked this out because of lack of additional information. Though, as much progress have been made recently in the open-source of fine-tuning softwares, I would advise patching the latest version of [trl](https://github.com/huggingface/trl) and apply weight interpolation on top of it.

Thank you for your prompt response. I will try it.
